### PR TITLE
core: add RunnableFactory class

### PR DESCRIPTION
**Pull request by skozlovf**
**Description:**
This PR adds a new `RunnableFactory` class that allows to create another runnable at runtime using a factory method.
Key features:
* Loadable parameters and validation thanks to ``pydantic``.
* Inheritance support and all other OOP features.
* Serialize only chain specific parameters.

Example usage:
```python
from langchain_core.output_parsers import StrOutputParser
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.runnables import Runnable, RunnableFactory
from langchain_openai import ChatOpenAI
from langchain_core.pydantic_v1 import validator

class MyRunnable(RunnableFactory):
    role: str = "You are a helpful AI assistant."
    prompt: str = "Tell me a joke about {topic}"
    model: str = "gpt-3.5-turbo"

    @validator("prompt")
    def validate_prompt(cls, v):
        if "joke" not in v:
            raise ValueError("You should only ask for jokes.")
        return v

    def _create_runnable(self) -> Runnable:
        prompt = ChatPromptTemplate.from_messages(
            [("system", self.role), ("human", self.prompt)]
        )
        llm = ChatOpenAI(model_name=self.model)

        chain = prompt | llm | StrOutputParser()
        return chain

# Create a chain with defaults.

chain = MyRunnable()
output = chain.invoke({"topic": "bear"})
print(output)

# Create a chain with custom parameters.

chain = MyRunnable(model="gpt-4-1106-preview")
output = chain.invoke({"topic": "bear"})
print(output)

# Create a chain from an external "source".

external_parameters = {
    "model": "gpt-4-1106-preview",
    "prompt": "Tell me two jokes about {topic}",
}

chain = MyRunnable.parse_obj(external_parameters)
#
# or
# chain = MyRunnable.parse_file(filename)
#
output = chain.invoke({"topic": "bear"})
print(output)

# Serialize parameters.
parameters = chain.dict()
chain = MyRunnable.parse_obj(parameters)
output = chain.invoke({"topic": "bear"})
print(output)
```

If you think this make sense, I can continue adding docs/tests.

### CODE REVIEW

**Comment by eyurtsev** [comment]
What are some examples where a RunnableFactory would work better than a function?

```python
def create_runnable(role, prompt, model):
        prompt = ChatPromptTemplate.from_messages(
            [("system", role), ("human", prompt)]
        )
        llm = ChatOpenAI(model_name=model)

        chain = prompt | llm | StrOutputParser()
        return chain

runnable  = create_runnable(...)
```

**Comment by skozlovf** [comment]
This is useful if you use OOP in your app.

If you use a plain class then flow is similar to the function way:
```python

class MyRunnable:
    role: str
    prompt: str
    model: BaseLanguageModel

    def create() -> Runnable:
        ...

chain = ...
factory = MyRunnable(...)
chain = chain | factory.create()
```
You can't put the class instance to a LCEL chain, you need a chain builder method that returns a `Runnable`.

`RunnableFactory` allows you to encapsulate your data and logic, and create custom runnable type.

This means that you can embed such instance to a LCEL chain directly, without extra actions:
```python
class MyRunnable(RunnableFactory):
    role: str
    prompt: str
    model: BaseLanguageModel

    def _create_runnable(self) -> Runnable:
        ...

chain = ...
my_chain = MyRunnable(...)
chain = chain | my_chain
```

Also, you can put it to a `pydantic` model field, and it will be automatically validated and created:
```python
class MyRunnable(RunnableFactory):
    role: str
    prompt: str
    model: BaseLanguageModel

    def _create_runnable(self) -> Runnable:
        ...

class AnotherRunnable(RunnableFactory):
    param: int
    my_chain: MyRunnable

    def _create_runnable(self) -> Runnable:
        chain = ...
        return chain | self.my_chain

chain = AnotherRunnable(...)
output = chain.invoke(...)
```

Side effect is what you can serialize *app-specific* parameters. `RunnableSerializable` allows to serialize already constructed chain. There is no documented way to load it, and furthermore, the serialized parameters are *chain-specific*:
```python

chain = AnotherRunnable.parse_obj(
    {
        "param": 2,
        "my_chain": {
            "role": "xxx",
            "prompt": "yyy",
        }
    }
)

params = chain.dict(...)
# {"param": 2, "my_chain": {"role": "xxx", "prompt": "yyy"}}
```

NOTE: I skipped `model` processing for simplicity.

PS Feel free to discard the PR if you don't think it should be included in the "core".

File: libs/core/langchain_core/runnables/factory.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        chain = MyRunnable.parse_obj(external_parameters)
+        output = chain.invoke({"topic": "bear"})
+        print(output)
+    """
+
+    _runnable: Runnable = PrivateAttr()
+
+    def __init__(self, **kwargs: Any) -> None:
+        super().__init__(**kwargs)
+        self._runnable = self._create_runnable()
```
Comment (baskaryan) [inline libs/core/langchain_core/runnables/factory.py:85]: if the runnable is created on init of the factory then it's not really a factory is it? you could've just initialized the underlying runnable, no?

Comment (skozlovf) [reply libs/core/langchain_core/runnables/factory.py:85]: That's right. The name is not very clear. Maybe `RunnableWrapper` or `RunnablePimpl` or smth else is better. One of the goals is to redirect API calls to the underlying runnable created at runtime. Let me know if it makes sense to put this whole idea in the core.

**Comment by skozlovf** [comment]
Ok, it seems this class is very specific for the `core`. I close the PR as there are more important LC tasks.

**Pull request closed without merging**